### PR TITLE
[esphome][ota] Disable loop while idle, wake on listening-socket activity

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

`ESPHomeOTAComponent::loop()` ran every main-loop tick just to check `client_ != nullptr || server_->ready()`. That looks cheap, but the `ready()` check reads `sock->rcvevent` as a `volatile int16_t` — which on Xtensa (ESP32) forces the compiler to emit a `memw` memory-wait instruction to serialize cross-CPU observability, then an `l16si` load, on **every idle tick forever**. Same story on the raw-TCP platforms: `ready()` touches volatile pointer fields that can't be cached. OTA is idle the vast majority of the device's life, so that's pure overhead paid to learn "still nothing."

OTA now disables its own loop after the first idle tick and the existing socket-wake paths re-enable it when a client actually connects. Zero `memw`'s, zero volatile loads, zero loop dispatches until there's real work to do.

## Cost

ESP32 / LibreTiny (fast select) sees a net RAM reduction in all measured configs — disabling the idle loop iteration removes OTA from the active dispatch partition and the 8 bytes of new BSS (`global_esphome_ota_component` + `s_ota_listener_conn`) are more than offset. ESP8266 (raw TCP, no fast-select static) pays +4 bytes RAM.

The entire feature is self-contained in the `esphome` OTA platform TU:

- **1 pointer (`global_esphome_ota_component`) in `ota_esphome.cpp`** — file-scope static, set in `setup()`, read by the extern `"C"` wake trampoline. Standard single-instance pattern already used throughout the codebase. Only linked in when `USE_OTA_PLATFORM_ESPHOME` is compiled in.
- **1 pointer (`s_ota_listener_conn`) in `lwip_fast_select.c`** — only on fast-select platforms (ESP32 IDF, BK72xx, RTL87xx). Stores the listener netconn so the RCVPLUS callback can filter wakes down to just OTA's listen socket.

## Wake hooks

This is deliberately OTA-only. OTA is in essentially every ESPHome config and has a uniquely idle profile (months of silence, then one inbound connection). A generic per-socket wake hook would bloat every monitored socket to serve the one caller that needs it.

- **ESP32 / LibreTiny (lwip fast select)** — `esphome_socket_event_callback` on `NETCONN_EVT_RCVPLUS`, filtered against `s_ota_listener_conn` so API client data, mDNS, etc. don't pay the wake cost. Calls an `extern "C" esphome_wake_ota_component_any_context()` trampoline defined in `ota_esphome.cpp`.
- **ESP8266 / RP2040 (raw TCP)** — `LWIPRawListenImpl::accept_fn_`, per-pcb via `tcp_accept()`. Implicitly filtered: the callback only fires on completed handshakes to the specific listener. Calls the same extern `"C"` trampoline directly.

Both paths gate on `USE_OTA_PLATFORM_ESPHOME` (added as a `-D` flag from `components/esphome/ota/__init__.py`) so the trampoline symbol only needs to exist when the esphome OTA platform is actually compiled in.

Runtime measurements from a real BK72xx device (via `runtime_stats`) are posted as a PR comment.

## Breaking change: single port for `ota: platform: esphome`

[#7001](https://github.com/esphome/esphome/pull/7001) added same-port merging for package overrides (preserved unchanged). It didn't cover two configs surviving on *different* ports, which silently became two independent listening sockets — never the feature's intent. `ota_esphome_final_validate` now raises `cv.Invalid` when more than one port survives merging. This also matches the single-instance invariant the wake pointer relies on.

Multi-instance `ota: platform: esphome` was never documented on [esphome.io/components/ota/esphome](https://esphome.io/components/ota/esphome/) — the page only ever showed a single entry — so no `esphome-docs` update is needed.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected) — [policy](https://developers.esphome.io/contributing/code/#what-constitutes-a-c-breaking-change)
- [ ] Developer breaking change (an API change that could break external components) — [policy](https://developers.esphome.io/contributing/code/#what-is-considered-public-c-api)
- [ ] Undocumented C++ API change (removal or change of undocumented public methods that lambda users may depend on) — [policy](https://developers.esphome.io/contributing/code/#c-user-expectations)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- n/a

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- n/a (no user-visible configuration change beyond the new validation error for the multi-port edge case)

## Test Environment

End-to-end OTA upload verified on real hardware across both wake paths:

- [ ] ESP32 (Arduino)
- [x] ESP32 IDF — BSD sockets + `USE_LWIP_FAST_SELECT`
- [x] ESP8266 — raw TCP
- [x] RP2040/RP2350 — raw TCP (low-priority user IRQ context)
- [x] BK72xx — LibreTiny LWIP sockets + `USE_LWIP_FAST_SELECT`
- [x] RTL87xx — LibreTiny LWIP sockets + `USE_LWIP_FAST_SELECT`
- [ ] LN882x
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# Common case — unchanged, behavior is transparent.
ota:
  - platform: esphome
    password: !secret ota_password
```

Still valid — same-port merge path from #7001 is unchanged:

```yaml
# User override of a package that already defined ota: - platform: esphome
ota:
  - platform: esphome
    password: !secret ota_password   # merged with the package's entry on the default port
```

Invalid after this PR (raises `cv.Invalid` at config time):

```yaml
ota:
  - platform: esphome
    port: 3232
  - platform: esphome
    port: 3233   # second port is no longer allowed
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/component_tests/ota/test_esphome_ota.py`).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/esphome/ota/ota_esphome.cpp [hunk 1]
```diff
@@ -65,6 +65,12 @@ void ESPHomeOTAComponent::setup() {
     this->server_failed_(LOG_STR("listen"));
     return;
   }
+
+  // Disable loop() while idle. Socket wake paths (LwIP fast select, raw TCP accept
+  // callback, host select) call App.wake_ota_component_any_context() which re-enables
+  // this component's loop when an incoming connection arrives.
+  App.set_ota_wake_component(this);
+  this->disable_loop();
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/esphome/ota/ota_esphome.cpp]: Addressed in 0f8419f: enforced single `ESPHomeOTAComponent` instance in `ota_esphome_final_validate` (raises `cv.Invalid` when more than one port survives merging) rather than growing the wake hook to support N components. Multi-port OTA was never a real use case — the port-based merge existed so a user config + remote package declaring OTA would merge instead of breaking, which still works.

File: esphome/core/application.h [hunk 2]
```diff
@@ -559,6 +559,20 @@ class Application {
...
+#ifdef USE_OTA
+  /// Register the OTA component so socket-wake paths can enable its loop when
+  /// a new connection arrives on the listening socket. OTA disables its own
+  /// loop while idle to avoid per-tick dispatch overhead.
+  void set_ota_wake_component(Component *component) { this->ota_wake_component_ = component; }
+  /// Wake the registered OTA component (if any) from any context. Safe to call
+  /// from the LwIP TCP/IP task and other callback contexts.
+  void IRAM_ATTR wake_ota_component_any_context() {
+    if (this->ota_wake_component_ != nullptr) {
+      this->ota_wake_component_->enable_loop_soon_any_context();
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/core/application.h]: Addressed in 0f8419f: `wake_ota_component_any_context()` now sets `pending_enable_loop_` and `has_pending_enable_loop_requests_` directly (Application is already a friend of Component) instead of calling `enable_loop_soon_any_context()`. Every caller has already woken the main loop by that point, so the second wake was redundant as flagged.

File: esphome/components/esphome/ota/ota_esphome.cpp [hunk 3]
```diff
@@ -566,6 +577,8 @@ void ESPHomeOTAComponent::cleanup_connection_() {
 #ifdef USE_OTA_PASSWORD
   this->cleanup_auth_();
 #endif
+  // Back to idle — sleep until the next incoming connection wakes us.
+  this->disable_loop();
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/esphome/ota/ota_esphome.cpp]: Addressed in af8fd1d: `cleanup_connection_()` no longer calls `disable_loop()`. `loop()` itself is idempotent — it re-checks `server_->ready()` on the next iteration and either accepts the queued client or disables cleanly. This avoids the race where `pending_enable_loop_` was set while the component was active (invisible to `enable_pending_loops_()`, which only scans the inactive section) and then lost when cleanup dropped the component to LOOP_DONE.

File: esphome/core/lwip_fast_select.c [hunk 4]
```diff
@@ -157,6 +157,13 @@ _Static_assert(offsetof(struct lwip_sock, rcvevent) == ESPHOME_LWIP_SOCK_RCVEVEN
 // Saved original event_callback pointer — written once in first hook_socket(), read from TCP/IP task.
 static netconn_callback s_original_callback = NULL;
 
+#ifdef USE_OTA
+// Extern wake hook for the OTA component (implemented in application.cpp). Called from the
+// TCP/IP task so the OTA component's disabled loop can be re-enabled when a new connection
+// arrives on its listening socket. Safe from task context via enable_loop_soon_any_context().
+extern void esphome_wake_ota_component_any_context(void);
+#endif
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/core/lwip_fast_select.c]: Fixed in af8fd1d: comment now calls out that `esphome_wake_ota_component_any_context()` fires on every RCVPLUS across all monitored sockets (so false wakes from unrelated sockets are expected) and that it only sets pending-enable flags — the `xTaskNotifyGive` above handles waking the main task.

File: esphome/components/socket/lwip_raw_tcp_impl.cpp [hunk 5]
```diff
@@ -856,6 +859,12 @@ err_t LWIPRawListenImpl::accept_fn_(struct tcp_pcb *newpcb, err_t err) {
   LWIP_LOG("Accepted connection, queue size: %d", this->accepted_socket_count_);
   // Wake the main loop immediately so it can accept the new connection.
   esphome::wake_loop_any_context();
+#ifdef USE_OTA
+  // Re-enable the OTA component loop if it disabled itself while idle.
+  // enable_loop_soon_any_context() is IRAM/IRQ-safe, which is required on RP2040
+  // where this callback runs in a low-priority user IRQ context.
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/socket/lwip_raw_tcp_impl.cpp]: Fixed in af8fd1d: comment rewritten to say the hook only sets pending-enable flags; the `wake_loop_any_context()` call above has already woken the main loop. Also notes the low-priority user IRQ context on RP2040 is safe because the hook only writes volatile bools.

File: esphome/core/application.cpp [hunk 6]
```diff
@@ -449,6 +449,12 @@ void Application::enable_pending_loops_() {
   }
 }
 
+#if defined(USE_OTA) && defined(USE_LWIP_FAST_SELECT)
+// Called from the LwIP TCP/IP task via esphome_socket_event_callback() on NETCONN_EVT_RCVPLUS.
+// enable_loop_soon_any_context() is task-safe and IRAM-resident.
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/core/application.cpp]: Fixed in af8fd1d: dropped the `enable_loop_soon_any_context()` reference; the shim now accurately describes forwarding to `App.wake_ota_component_any_context()`, which only marks the pending-enable flags (the fast-select callback has already woken the main task via `xTaskNotifyGive`).

File: esphome/components/esphome/ota/__init__.py [hunk 7]
```diff
@@ -78,6 +78,14 @@ def ota_esphome_final_validate(config):
         else:
             new_ota_conf.append(ota_conf)
 
+    if len(merged_ota_esphome_configs_by_port) > 1:
+        raise cv.Invalid(
+            f"Only a single '{CONF_OTA}' '{CONF_PLATFORM}: {CONF_ESPHOME}' instance is "
+            f"supported, but multiple were configured on different ports "
+            f"({sorted(merged_ota_esphome_configs_by_port.keys())}). Remove the extra "
+            f"configurations or place them on the same port so they can be merged."
+        )
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/esphome/ota/__init__.py:87]: Added in af8fd1d: `tests/component_tests/ota/test_esphome_ota.py` covers single instance accepted, same-port configs merge cleanly, different-port configs rejected with `cv.Invalid`, and non-esphome OTA platforms unaffected.

File: esphome/components/esphome/ota/ota_esphome.cpp [hunk 8]
```diff
@@ -81,13 +87,18 @@ void ESPHomeOTAComponent::dump_config() {
...
-  // This optimization reduces idle loop overhead when OTA is not active
-  // Note: No need to check server_ for null as the component is marked failed in setup()
-  // if server_ creation fails
-  if (this->client_ != nullptr || this->server_->ready()) {
-    this->handle_handshake_();
+  // loop() is disabled while idle (see setup() / cleanup_connection_()). Socket-wake
+  // paths (LwIP fast select, raw TCP accept, host select) call
+  // App.wake_ota_component_any_context() to re-enable this loop when a monitored
+  // socket signals activity. False wakes (e.g. an API-socket event) land here with
+  // no pending work — in that case we disable the loop again and go back to sleep.
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/esphome/ota/ota_esphome.cpp]: Fixed in 903a159: `loop()` docstring now says loop starts disabled in `setup()` and explicitly notes that `cleanup_connection_()` intentionally does NOT disable the loop (so one more iteration runs to catch any connection queued during the session).

File: esphome/core/lwip_fast_select.h [hunk 9]
```diff
@@ -53,6 +53,14 @@ static inline bool esphome_lwip_socket_has_data(struct lwip_sock *sock) {
 /// The sock pointer must have been obtained from esphome_lwip_get_sock().
 void esphome_lwip_hook_socket(struct lwip_sock *sock);
 
+/// Filter the inline OTA wake hook in the fast-select callback so it only fires for
+/// RCVPLUS events on this specific listener's netconn. Without this, every monitored
+/// socket's RCVPLUS (API client data, web server, mDNS, etc.) would mark OTA
+/// pending-enable and force its loop to run a wake-up tick only to re-disable itself.
+/// Captured at OTA setup(); stays pointing at the listener for the device's lifetime.
+/// Pass NULL to clear the filter (wake fires on every RCVPLUS, pre-filter behavior).
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/core/lwip_fast_select.h]: Fixed in f75d6ab: rewrote the header comment to describe the actual semantics — passing NULL disables OTA wakes entirely (since `conn == NULL` never matches a real event), not "wake fires on every RCVPLUS." That NULL-is-off behavior is what we want before the listener is installed and after it's torn down.

File: esphome/components/esphome/ota/ota_esphome.cpp [hunk 10]
```diff
@@ -81,13 +98,28 @@ void ESPHomeOTAComponent::dump_config() {
...
+  // Self-disabling idle loop. On the first tick after setup() (and after every session
+  // cleanup and every false wake), if there's no client and the listener has nothing
+  // queued, we disable ourselves and go back to sleep. Socket-wake paths (LwIP fast
+  // select, raw TCP accept, host select) mark us pending-enable via
+  // App.wake_ota_component_any_context() when a monitored socket signals activity, and
+  // enable_pending_loops_() reactivates us.
+  //
+  // False wakes from unrelated monitored sockets are expected — the event callbacks
+  // fire on every RCVPLUS across all monitored sockets, not just OTA's listener — and
+  // they land here with no pending work.
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/esphome/ota/ota_esphome.cpp]: Fixed in f75d6ab: rewrote the loop() docstring to describe post-filter behavior. Fast-select is filtered to OTA's listener netconn, and raw TCP's accept_fn_ is inherently per-pcb, so unrelated monitored sockets no longer wake OTA through either path. The idle self-disable is retained as a safety net for the narrow cases where a wake can land with no pending work (connection queued during an active session, filter not yet installed, host select fallback).

**Comment by bdraco** [comment]
## Runtime proof

Measured on a real BK72xx (cb7231n) device running this branch, via the `runtime_stats` component. OTA is registered, network is up, Home Assistant API client is connected, device is otherwise idle.

**First 60 seconds after boot** — OTA's `loop()` runs exactly once to self-disable on its first idle tick:

```
Period stats (last 60000ms): 11 active components
  api: count=7240, avg=0.008ms, max=10.53ms, total=60.8ms
  safe_mode: count=7240, ...
  ...
  esphome.ota: count=1, avg=0.022ms, max=0.02ms, total=0.0ms
```

**Next 60 seconds** — steady state, API still connected and busy:

```
Period stats (last 60000ms): 5 active components
  api: count=7480, ...
  preferences, uptime.sensor, template.sensor, web_server
```

`esphome.ota` is **not in the list at all.** Zero `loop()` invocations in the 60-second window — it stays disabled indefinitely. The wake hook only fires on real connection attempts to the OTA listen socket, and there haven't been any, so the component sits completely idle.

**Third 60-second window** — still zero OTA invocations:

```
Period stats (last 60000ms): 5 active components
  api: count=7492, ...
```

Before this PR, the same device would have shown `esphome.ota: count≈3750` in every period for its entire lifetime — every ~16 ms loop tick paying a `memw` + volatile `int16_t` load (the `rcvevent` read inside `server_->ready()`) just to learn "still nothing." This PR eliminates all of it.

File: esphome/components/esphome/ota/ota_esphome.cpp [hunk 11]
```diff
@@ -28,6 +31,17 @@ static constexpr size_t OTA_BUFFER_SIZE = 1024;                  // buffer size
...
 static constexpr uint32_t OTA_SOCKET_TIMEOUT_DATA = 90000;       // milliseconds for data transfer
 
+// Single-instance pointer — multi-port configs are rejected in final_validate.
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+static ESPHomeOTAComponent *global_esphome_ota_component = nullptr;
+
+// Called from any context (LwIP TCP/IP task, RP2040 user-IRQ).
+extern "C" void esphome_wake_ota_component_any_context() {
+  if (global_esphome_ota_component != nullptr) {
+    global_esphome_ota_component->enable_loop_soon_any_context();
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/esphome/ota/ota_esphome.cpp:41]: its protected and probably not worth adding another public api to avoid one extra wake on a component that would otherwise run 3,679,200,000/loops a year.

[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/esphome/ota/ota_esphome.cpp:41]: The redundant wake is real but negligible: it fires once per actual OTA connection (not per idle tick), so it's an extra idempotent `xTaskNotifyGive` on the order of microseconds maybe a few times a year in practice. Avoiding it cleanly would require adding a new protected `Component` helper to set the flags without the wake, since `Component::pending_enable_loop_` and `Application::has_pending_enable_loop_requests_` are both protected and friendship doesn't extend to the extern "C" trampoline. Not worth a new API on a component that would otherwise loop ~3.7 billion times/year.

**Review by swoboda1337 — APPROVED** [review]

**Comment by bdraco** [comment]
thanks
